### PR TITLE
YJIT: add support for calling bmethods

### PR DESCRIPTION
**Pull request by XrXr**
This commit adds support for the VM_METHOD_TYPE_BMETHOD method type in
YJIT. You can get these type of methods from facilities like
Kernel#define_singleton_method and Module#define_method.

Even though the body of these methods are blocks, the parameter setup
for them is exactly the same as VM_METHOD_TYPE_ISEQ, so we can reuse
the same logic in gen_send_iseq(). You can see this from how
vm_call_bmethod() eventually calls setup_parameters_complex() with
arg_setup_method.

Bmethods do need their frame environment to be setup differently. We
handle this by allowing callers of gen_send_iseq() to control the iseq,
the frame flag, and the prev_ep. The `prev_ep` goes into the same
location as the block handler would go into in an iseq method frame.

---

By the way, I also have a PR to add a fast path for bmethods in the interpreter: #6428

### CODE REVIEW

File: yjit/src/codegen.rs [hunk 1]
```diff
@@ -4443,17 +4451,59 @@ fn push_splat_args(required_args: i32, ctx: &mut Context, asm: &mut Assembler, o
...
+
+    // Passing a block to a block needs logic different from passing
+    // a block to a method and sometimes requires allocation. Bail for now.
+    if block.is_some() {
+        gen_counter_incr!(asm, send_bmethod_block_arg);
+        return CantCompile;
+    }
+
+    let frame_type = VM_FRAME_MAGIC_BLOCK | VM_FRAME_FLAG_BMETHOD | VM_FRAME_FLAG_LAMBDA;
+    gen_send_iseq(jit, ctx, asm, ocb, iseq, ci, frame_type, Some(capture.ep), cme, block, argc)
```
Comment (jhawthorn) [inline yjit/src/codegen.rs:4491]: 😍

**Comment by noahgibbs** [comment]
Nice! This is a very clean way to implement this.

**Review by maximecb — APPROVED** [review]
Wonderful, thanks for adding tests too! :)

**Comment by jhawthorn** [comment]
It looks like this took us on railsbench (based on CI) from

```
ratio_in_yjit:              86.0%
```

to

```
ratio_in_yjit:              87.4%
```

**Pull request merged**